fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBJEJournal.java

```diff
@@ -390,6 +409,91 @@ public List<Long> getDatabaseNames() {
         return bdbEnvironment.getDatabaseNames();
     }
 
+    @Override
+    public void writeWithinTxn(short op, DataOutputBuffer buffer) {
+        if (currentTrasaction == null) {
+            staggingNextJournalId = nextJournalId.get();
+            stagingStartJournalId = staggingNextJournalId;
+        }
+        // id is the key
+        DatabaseEntry theKey = new DatabaseEntry();
+        TupleBinding<Long> idBinding = TupleBinding.getPrimitiveBinding(Long.class);
+        idBinding.objectToEntry(staggingNextJournalId, theKey);
+
+        // entity is the value
+        DatabaseEntry theData = new DatabaseEntry(buffer.getData(), 0, buffer.getLength());
+        if (MetricRepo.isInit) {
+            MetricRepo.COUNTER_EDIT_LOG_SIZE_BYTES.increase((long) theData.getSize());
+        }
+        LOG.debug("opCode = {}, journal size = {}", op, theData.getSize());
+
+
+        for (int i = 0; i < RETRY_TIME; i++) {
+            try {
+                // sleep before retry
+                if (i != 0) {
+                    Thread.sleep(SLEEP_INTERVAL_SEC * 1000);
+                }
+
+                // start txn
+                if (currentTrasaction == null) {
+                    currentTrasaction = currentJournalDB.getDb().getEnvironment().beginTransaction(
+                            null, bdbEnvironment.getTxnConfig());
+                }
+
+                // put, not commit
+                if (currentJournalDB.put(currentTrasaction, theKey, theData) == OperationStatus.SUCCESS) {
+                    LOG.debug("write journal {} within txn {} finished. db name {}",
+                            staggingNextJournalId, currentTrasaction.getId(), currentJournalDB.getDb().getDatabaseName());
+                    staggingNextJournalId += 1;
+                    return;
+                }
+            } catch (DatabaseException | InterruptedException e) {
+                LOG.error("catch an exception when writing to database. sleep and retry. journal id {}, txn {}",
+                        staggingNextJournalId, currentTrasaction, e);
+            }
+        }
+
+        // fail even after retry
+        String msg = "write bdb within txn failed. will exit. journalId: " + staggingNextJournalId + ", txnId: " +
+                currentTrasaction.getId() + ", bdb database Name: " + currentJournalDB.getDb().getDatabaseName();
+        LOG.error(msg);
+        Util.stdoutWithTime(msg);
+        System.exit(-1);
+    }
+
+    @Override
+    public void commitTxn() {
+        assert (currentTrasaction != null);
```

Comment:
If commitTxn() failed but writeWithinTxn() succeed, is there necessary to rollbock the BDB?

Reply:
No, nothing is written if commit failed.